### PR TITLE
Implement HLE D3DDevice_RunVertexStateShader()

### PR DESCRIPTION
**Pull request by jackchentwkh**
Thanks for @RadWolfie 's help, I can clean up my messed up codes. 
here is a PR with better and clear commits.
if this PR got merged, I will close the other PR.

### CODE REVIEW

**Comment by jackchentwkh** [comment]
this PR is supposed to replace https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/pull/2380

**Review by PatrickvL — COMMENTED** [review]
One suggestion

**Comment by jackchentwkh** [comment]
@PatrickvL you mean like this?
memset(vertex_state_shader_v0, 0, 16);

**Comment by RadWolfie** [comment]
> This can be replaced by `memset(vertex_state_shader_v0, 0, 4 * size of(float))`

I wouldn't advice to use this suggestion. Instead, it should be `std::memset(vertex_state_shader_v0, 0, sizeof(vertex_state_shader_v0));`. sizeof macro count both float and array as bytes size. If need to count array size, then it will be `ARRAY_SIZE(vertex_state_shader_v0)` as part of our own std extension.

File: src/core/hle/D3D8/Direct3D9/Direct3D9.cpp [hunk 1]
```diff
@@ -8676,8 +8681,38 @@ xbox::void_xt WINAPI xbox::EMUPATCH(D3DDevice_RunVertexStateShader)
...
 	// If pData is assigned, pData[0..3] is pushed towards nv2a transform data registers
 	// then sends the nv2a a command to launch the vertex shader function located at Address
+	NV2AState* dev = g_NV2A->GetDeviceState();
+	PGRAPHState* pg = &(dev->pgraph);
 
-    LOG_UNIMPLEMENTED(); 
+	float vertex_state_shader_v0[4];
+	if (pData != nullptr)
+		//if pData != nullptr, then it contents v0.xyzw, we shall copy the binary content directly.
+		memcpy(vertex_state_shader_v0,pData, sizeof(vertex_state_shader_v0));
```
Comment (RadWolfie) [inline src/core/hle/D3D8/Direct3D9/Direct3D9.cpp:8690]: One small nit, add a space between comma and pData.

**Comment by RadWolfie** [comment]
I have one suggestion for you to do locally, perform a rebase on top of master branch to clear the merge commit.

**Comment by jackchentwkh** [comment]
titles known using RunVertexStateShader()
SVS : fixed
NBA Jam : not tested yet.
Robotech Invasion : fixed
Curious George : not tested yet. 
Tak 2 The Staff Of Dreams : don't know the previous status, but in this PR it's playable.
Crash Twinsanity : previsouly it's already playable, I didn't see any thing improved obviously.

known regression:
Halo: won't get in game after menu.

**Comment by medievil1** [comment]
as I said on discord, Halo WILL go ingame, but it is back to being hit and miss and takes a whole bunch of tries...there was an audio fix that made it more consistent before, now I think it is more timing...several games have some timing issues especially on videos (which immediately after menu, halo loads a small loading video which does not display currently), far cry for instance stops and starts the logo video (CPU usage on 1 core hits 100% as well)...so I lean toward threading issue for halo

**Comment by Triticum0** [comment]
@jackchentwkh Most of the issues on Xemu 309, were mostly to do with clipping issues. The game on your list that uses RunVertexStateShader() is the only game to my knowledge which use it. Other than the outlier NBA Jam All the games that use it use the Vicious Engine post-2003

**Comment by jackchentwkh** [comment]
abaire, the author of nv2a_vsh_cpu sub module and the xemu NV097_LAUNCH_TSLRANSFORM_PROGRAM implementation PR, agrees with my usage of his code in cxbx-reloaded project.

**Comment by RadWolfie** [comment]
I tried my best to find the titles, instead of looking up each one from the list. I went through SymbolCache folder to find titles use D3DDevice_RunVertexStateShader function, I do not see any improvement nor regression. I'm not entirely sure, but Silent Hill 2 may fall under clipping issue category for character isn't visible and can partially see texture stretching around it.

**Comment by medievil1** [comment]
I can confirm wreckless and metal slug 3 both work now......

**Review by LukeUsher — APPROVED** [review]
Excellent work, thank you.

**Pull request merged**